File: algorithms/ar-lscsa/typescript/largest_sum_contiguous_subarray.ts
function largestSumContiguousSubarray (arr: number[]) {
    let globalMax = Number.MIN_VALUE;
    let localMax = 0;

    for (const num of arr) {
        localMax += num;

        if (globalMax < localMax) {
            globalMax = localMax;
        }
        if (localMax < 0) {
            localMax = 0;
        }
    }

    return globalMax;
}

function main () {
    let input = [2, 3, 0, 4];
    largestSumContiguousSubarray(input);
}

main();

File: algorithms/ar-lscsa/csharp/LargestSumContiguousSubArray.cs
using System;

public class LargestSumContiguousSubArray
{
    public static int DoLargestSumContiguousSubArray(int[] arr)
    {
        int max = int.MinValue, maxEnd = 0;

        foreach (int num in arr)
        {
            maxEnd = maxEnd + num;

            if (max < maxEnd) {
                max = maxEnd;
            }

            if (maxEnd < 0) {
                maxEnd = 0;
            }
        }

        return max;
    }

    public static void Main()
    {
        int[] arr = new int[] {-2, 1, -3, 4, -1, 2, 1, -5, 4};
        int result = DoLargestSumContiguousSubArray(arr);

        Console.WriteLine(result);
    }
}
